src/Tools/dotnet-monitor/CollectionRules/Actions/CollectStacksAction.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Exceptions;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options.Actions;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Threading;
+using System.Threading.Tasks;
+using Utils = Microsoft.Diagnostics.Monitoring.WebApi.Utilities;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions
+{
+    internal sealed class CollectStacksActionFactory : ICollectionRuleActionFactory<CollectStacksOptions>
+    {
+        private readonly IServiceProvider _serviceProvider;
+
+        public CollectStacksActionFactory(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+        }
+
+        public ICollectionRuleAction Create(IEndpointInfo endpointInfo, CollectStacksOptions options)
+        {
+            if (null == options)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            ValidationContext context = new(options, _serviceProvider, items: null);
+            Validator.ValidateObject(options, context, validateAllProperties: true);
+
+            return new CollectStacksAction(_serviceProvider, endpointInfo, options);
+        }
+    }
+
+    internal sealed class CollectStacksAction : CollectionRuleActionBase<CollectStacksOptions>
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ProfilerChannel _profilerChannel;
+
+        public CollectStacksAction(IServiceProvider serviceProvider, IEndpointInfo endpointInfo, CollectStacksOptions options) : base(endpointInfo, options)
+        {
+            _serviceProvider = serviceProvider;
+            _profilerChannel = _serviceProvider.GetRequiredService<ProfilerChannel>();
+        }
+
+        protected override async Task<CollectionRuleActionResult> ExecuteCoreAsync(TaskCompletionSource<object> startCompletionSource, CollectionRuleMetadata collectionRuleMetadata, CancellationToken token)
+        {
+            bool isPlainText = Options.Format.GetValueOrDefault(CollectStacksOptionsDefaults.DefaultFormat) == CallStackFormat.PlainText;
```

Comment:
nit: can you add an extension method for this? `CollectStacksOptionsExtensions.GetFormat`